### PR TITLE
feat(ctb): Account for worst-case `CALL` gas in `SafeCall`

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adjusts `SafeCall.callWithMinGas` to account for the worst-case cost of 3/5 of the `CALL` opcode's dynamic gas factors:
* `address_access_cost`
* `positive_value_cost`
* `value_to_empty_account_cost`

As for `memory_expansion_cost` & `code_execution_cost`, these are expected to be covered by the `minGasLimit` itself.

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5470** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/monorepo/5470" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/monorepo/5470?utm_source=stack-comment).

File: packages/contracts-bedrock/contracts/test/invariants/SafeCall.t.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/test/invariants/SafeCall.t.sol:104]: Should be able to access this as `CONSOLE`
https://github.com/foundry-rs/forge-std/blob/fc560fa34fa12a335a50c35d92e55a6628ca467c/src/Base.sol#L11

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:397]: nit: prefer kwargs syntax when its unclear what the arguments have become, ie the magic `0`

File: packages/contracts-bedrock/contracts/test/invariants/SafeCall.t.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/test/invariants/SafeCall.t.sol:104]: `CONSOLE` is a part of `Base.sol`, and this actor contract only inherits from `StdUtils`.

**Comment by maurelian** [comment]
Fixes CLI-3832

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: I propose that we: 

1. Revert #5108. 
2. Rebase on #5444 once it's in.

Combine the old gas buffer check with the new reentrancy protection check to be:

```diff
    if (xDomainMsgSender != Constants.DEFAULT_L2_SENDER 
+	|| gasleft() >= _minGasLimit + RELAY_GAS_REQUIRED
    ) {
            failedMessages[versionedHash] = true;
            emit FailedRelayedMessage(versionedHash);
            return;
        }
```

Comment (tynes) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: I do like this direction but am unsure of exactly how it would help to unstick a message as we do not allow for passing in a different `minGasLimit`. The old contracts do allow for this, there is a different function `replayMessage`. This at least would give us better monitoring into the issue, we would be able to observe the set of failed messages

Comment (tynes) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: by `RELAY_GAS_REQUIRED` do you mean `MIN_GAS_CONSTANT_OVERHEAD`?

Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: > ... we do not allow for passing in a different `minGasLimit`

We don't allow modifying `minGasLimit`, but `SafeCall.call()` can accept `gasleft()` as its `_gas` [argument](https://github.com/ethereum-optimism/optimism/blob/d460e7670923a115fe2c3c3af14914df8394b849/packages/contracts-bedrock/contracts/libraries/SafeCall.sol#L17-L26) which it will forward.

`RELAY_GAS_REQUIRED` was the constant value in a previous version of the contract which was "the amount of gas held in reserve to guarantee that relay execution completes".

https://github.com/ethereum-optimism/optimism/blob/608ebc9a8a597d57acc71d4740fe9b1980357380/packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol#L370-L377

There were other issues with the implementation (not accounting for the SSTORE to set `xDomainMessageSender`), but we can address that.

Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: I think the pattern we need to avoid is any logic that triggers a revert based on the `gasleft()`.

Comment (clabby) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol]: I agree that not reverting here is the safer route. I've rebased the PR on top of the other XDM changes and changed it up.

File: packages/contracts-bedrock/contracts/L2/L2CrossDomainMessenger.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/L2/L2CrossDomainMessenger.sol:25]: I just realized that we forgot to bump semver on #5444. Do we need to bump this up twice to `1.3.0` to account for that? 

@tynes I think you have the clearest ideas about that.

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:155]: Need to reset the patch version when the minor version increments:

```suggestion
    ) Semver(1, 4, 0) {
```

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (refcell) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:361]: Not a big deal, so just a comment but I think instead of "silently fail", it'd be better verbiage to say something to the tune of "return early" since silently failing implies a resulting unexpected state or state to be out of wack.

File: packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol
Comment (tynes) [reply packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol:29]: Let's bump first outside of this PR and then rebase this on top of that

Comment (maurelian) [reply packages/contracts-bedrock/contracts/L1/L1CrossDomainMessenger.sol:29]: resolved #5491

**Review by refcell — APPROVED** [review]
Looks good to me :handshake:

**Comment by clabby** [comment]
Adding the `mergify-ignore` label so that this doesn't get merged in before we're confident that these changes plug all necessary holes.

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:393]: Would the following be slightly more accurate?

```suggestion
        //   2. The amount of gas provided to the execution context of the the target contract is at
        //       least the gas limit specified by the user. If there is not enough gas in the
        //       callframe to accomplish this, `callWithMinGas` will revert.
```

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:395]: It's slightly unclear to me if this is meant to indicate that this is handled by `SafeCall.callWithMinGas` or if it's just friendly reminder about how the EVM works. 

Also I don't think it'd be obvious to an outsider what the purpose of this statement is. 

Maybe it would help to append it with: 

```
This ensure that calls which do not provide sufficient additional gas will not finalize
the withdrawal, so that it can be attempted again.
```

Honestly maybe I'm just being pedantic and there's a lot of words here already, so I'm not going to block over this.

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:69]: I think this is a bit clearer

```suggestion
            // still possible to self-rekt by initiating a withdrawal with a minimum gas limit
            // that does not account for the `memory_expansion_cost + code_execution_cost` factors
            // of the dynamic cost of the `CALL` opcode.
```

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: I'm slightly worried about this breaking gas estimation

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/L1/OptimismPortal.sol]: Added this with a slight modification, this is more clear than the current comment imo :100:

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol]: Agree, added

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:395]: This comment was removed because the `callWithMinGas` function will **not** revert if there is not enough gas remaining to complete the rest of `finalizeWithdrawalTransaction` after the call is performed. In the case that there's not enough gas to complete execution, `finalizeWithdrawalTransaction` will OOG, allowing for the caller to retry with a higher gaslimit.

Comment (clabby) [reply packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:395]: This comment was also wrong prior to this change - The behavior of `callWithMinGas` has never attempted to reserve any more gas than was required to forward `_minGasLimit` to the target context.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:366]: Yeah, this totally breaks gas estimation, good spot. Adding a revert if the `tx.origin` is the estimation address.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (tynes) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol]: I am starting to feel reluctant about how large of a diff this is

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (smartcontracts) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: This doesn't actually reserve the gas amount

Comment (smartcontracts) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: Still passing in the full gas amount, the gas buffer needs to be reserved explicitly (gasleft() - relay gas buffer). I think best to just use callWithMinGas here and add a new parameter for reservedGas. It's much clearer.

Comment (clabby) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: The reason I didn't fully reserve the remaining gas was because the `RELAY_RESERVED_GAS` constant factors in both the worst case cost of the `CALL` opcode (minus the memory expansion / execution costs) as well as the gas required to finish the execution. We could split it up into 2 constants?

fwiw, I initially went w/ modifying `callWithMinGas`, but @maurelian brought up a great point that we can just not revert in this context if a threshold of remaining gas is not met, which is much safer. See the thread above in this PR ^

File: packages/contracts-bedrock/contracts/L1/OptimismPortal.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/L1/OptimismPortal.sol:395]: OMG, I was reviewing that comment as if it was an addition. 🫠

File: packages/contracts-bedrock/contracts/test/invariants/CrossDomainMessenger.t.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/test/invariants/CrossDomainMessenger.t.sol:103]: why `init`?

Comment (clabby) [reply packages/contracts-bedrock/contracts/test/invariants/CrossDomainMessenger.t.sol:103]:
```solidity
contract MultipleSetupTest is Test {
    function setUp() public {
        // ...
    }

    function setUp(bool _in) public {
        // ...
    }

    function testExample() public {
        // ...
    }
}
```

```
$ forge test

Running 3 tests for test/Sandbox.t.sol:ChallengeTest
Failing tests:
Encountered 1 failing test in test/Sandbox.t.sol:MultipleSetupTest
[FAIL. Reason: Multiple setUp functions] setUp() (gas: 0)
```

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:378]: Hmmm, I'm not sure this will work as the Portal's `finalizeWithdrawalTransaction` won't bubble it up.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: > See the thread above in this PR ^

@smartcontracts the thread is gone from this view, please take a look at it [here](https://github.com/ethereum-optimism/optimism/pull/5470#discussion_r1169217738).

Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: > We could split it up into 2 constants?

I believe it was that way previously and that it makes the most sense, at least to my mind.

File: packages/contracts-bedrock/contracts/test/invariants/CrossDomainMessenger.t.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/test/invariants/CrossDomainMessenger.t.sol:103]: okee dokee. :)

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (smartcontracts) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: Replacing `callWithMinGas` with `call` just has the effect of not explicitly guaranteeing that the sub-call will receive enough gas. So the minimum gas limit is no longer respected. I feel like this is trying to make a larger diff than the original intention which should be like a 10 line diff (add two buffers, make sure it always respects minimum gas limit).

I do understand that minimum gas limit is a pain here and we should consider modifying things in the future but I feel like this is a large unnecessary diff at this point

Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol:385]: > Replacing `callWithMinGas` with `call` just has the effect of not explicitly guaranteeing that the sub-call will receive enough gas.

It's true that `callWithMinGas` explicitly guarantees that the sub-call will receive enough gas, but it enforces that guarantee by bricking the withdrawal.

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:70]: This comment should be on `checkMinGas` above.

Some or all of the rest comment block below also needs to be moved or edited to acommodate this change.

Comment (clabby) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:70]: Still in the process of splitting these - will do.

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:45]: Nit:
> returns `true` if there is not enough gas remaining

This is unintuitive to me. For example, if I just encountered this function in the wild (with a less descriptive name for the return value), ie:

```
bool ret = checkMinGas(_minGas, 0)
```
I would assume that `ret == true` means we _do_ have enough gas.

Comment (clabby) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:45]: The parameter is named `_halt` for this reason - glad to add some extra documentation here or flip it around.

Comment (maurelian) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:45]: How about: 
- Flip it
- rename to `hasMinGas()`

Comment (maurelian) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:45]: Another nit: we've always avoided named vars. We should reenable that slither check.

```suggestion
    function checkMinGas(uint256 _minGas, uint256 _reservedGas) internal view returns (bool) {
```

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (clabby) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:92]: Comments updated, RFC

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (refcell) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol]: Think we're in a good place to resolve this convo

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (maurelian) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:120]: Comment only:

It's interesting to me that we give `gas()` (or rather `min(gas() * 63 / 64, gas())`) to the actual `CALL`, rather than `_minGas`, which in the Portal is `tx.gasLimit`.

https://github.com/ethereum-optimism/optimism/blob/0470651b3c425a74f9ba24113f0a695645e8a7f9/packages/contracts-bedrock/contracts/L1/OptimismPortal.sol#L394

It seems fine since we do all the work above to ensure that when we get here, `gas()` is going to be plenty higher `tx.gasLimit`. 

Also, for whatever reason `CALL` doesn't seem to fail if `gas()` is less then the stack value consumed by gas.

Comment (refcell) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:120]: I think it makes sense to pass any additional gas a caller provides into the subcontext since the parameter is specifically annotated as a _minimum_ gas limit.

**Review by refcell — APPROVED** [review]
This looks pretty ironed out to me at this point pending other's reviews.

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:120]: Yeah, I agree. Just needed to think it through.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol]: > Think we're in a good place to resolve this convo

Oh, didn't realize this was my comment to begin with. Resolved.

File: packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol
Comment (maurelian) [reply packages/contracts-bedrock/contracts/universal/CrossDomainMessenger.sol]: Resolving on the basis that this applies to outdated code.

**Review by maurelian — APPROVED** [review]

**Comment by clabby** [comment]
Waiting for @tynes' approval to remove the `do-not-merge` and `mergify-ignore` labels.

**Comment by tynes** [comment]
Do you mind squashing this PR in a commit or a few commits?

File: packages/contracts-bedrock/contracts/libraries/SafeCall.sol
Comment (tynes) [inline packages/contracts-bedrock/contracts/libraries/SafeCall.sol:67]: could you get rid of `iszero` by using `gt` instead of `lt`?

Comment (clabby) [reply packages/contracts-bedrock/contracts/libraries/SafeCall.sol:67]: Mentioned in slack as well, x-posting:
```
iszero(lt(x, y)) == x >= y
gt(x, y) == x > y
```

The off-by-one here isn't a big deal, but imo should stick to the `>=` assertion (6 gas + 3-6 gas for dups isn't a big deal w/ the giant buffers)

**Review by tynes — APPROVED** [review]

**Pull request merged**
